### PR TITLE
fix: Back press handling of GroupWise Fragment

### LINKED ISSUE #1593: GroupWiseSkillsFragment not loaded on pressing back button
**Actual Behaviour**

The GroupWiseSkillsFragment should be loaded when the user presses back button from the SkillDetailsFragments. (in case of group wise skills).

**Expected Behaviour**

The GroupWiseSkillsFragment should be loaded when the user presses back button. (in case of group wise skills).

**Steps to reproduce it**

Open the skills listing page. Scroll to the groups at the bottom. Click on a group. This will show all the skills belonging to a particular group. Click on any skills to load skills details. Now, press the back button. You will see the skills listing page and not the group wise skills list.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by NileshPranami**
Fixes #1593 

- Handle back press correctly in concerned places

### CODE REVIEW

**Comment by NileshPranami** [comment]
@iamareebjamal please review again my changes

**Comment by iamareebjamal** [comment]
I think the logic is not very good and will fail in future

**Comment by NileshPranami** [comment]
@iamareebjamal  Any suggestions regard this problem. It will help me a lot.

**Comment by iamareebjamal** [comment]
Set the title in respective onResume of fragments

**Comment by NileshPranami** [comment]
@iamareebjamal On pressing back button onResume() will not call so we can't set title into it. Is there any other method which called explicitly on pressing back button?

**Comment by iamareebjamal** [comment]
onResume is called on back press

**Comment by NileshPranami** [comment]
@iamareebjamal I have checked using log messages but it respond only when I have opened that particular fragment but on back press no one respond.

**Comment by iamareebjamal** [comment]
Try onStart then

**Comment by NileshPranami** [comment]
@iamareebjamal onStart and onResume are called after pressing back button only when Fragments are replaced instead of add. Because of using replace app is taking extra time on back.

**Comment by iamareebjamal** [comment]
Then you can add a FragmentTransaction listener solution

**Comment by NileshPranami** [comment]
@iamareebjamal  I have googled a lot but I didn't get listener for FragmentTransaction but listener for onBackStackChanged can do same thing but it doesn't make difference from onBackPressed.

 Otherwise we can make a separate method in SkillsActivity for set title to fragments on back pressed. 
And we can call this method from onBackPressed. Because in future if any changes will happen then it may help us.

screenshot 44

**Comment by iamareebjamal** [comment]
We don't want to have if else statements for setting titles

**Comment by NileshPranami** [comment]
@iamareebjamal  I have simplified the code as you told. Thanks for your responses. Those are much valuable to me.

**Comment by NileshPranami** [comment]
@iamareebjamal Sir is any further improvement required in codes?

**Comment by iamareebjamal** [comment]
Except onResume, why is any other code needed?

**Comment by NileshPranami** [comment]
@iamareebjamal  Sir here are explanations 
```
private fun exitActivity(context: Context) {
var backStackEntryCount=supportFragmentManager.backStackEntryCount      
        supportFragmentManager.popBackStack()                                                    
        if(backStackEntryCount==1) {                                                                      
            finish()
            //previous code                                                                            
       }
}
```
In above part I have added code to check whether app is at SkillListingFragment or not. In that stage on back press ChatActivity should be open.
```
supportFragmentManager.addOnBackStackChangedListener {
                val currentFragment: android.support.v4.app.Fragment? = supportFragmentManager.findFragmentById(R.id.fragment_container)
                currentFragment?.onResume()
            }
```
In above part I have added code to call onResume() on back pressed.

**Comment by iamareebjamal** [comment]
Actually, pop backstack is automatically done in fragment activity. The existing implementation is flawed itself. So I'm giving a go ahead

File: app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt [hunk 1]
```diff
@@ -64,21 +64,23 @@ class SkillsActivity : AppCompatActivity(), SkillFragmentCallback {
     }
 
     private fun exitActivity(context: Context) {
-        overridePendingTransition(R.anim.trans_right_in, R.anim.trans_right_out)
-        val intent = Intent(context, ChatActivity::class.java)
-        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
-        startActivity(intent)
+        var lastFragment: android.support.v4.app.Fragment? = supportFragmentManager.findFragmentById(R.id.fragment_container)
+        supportFragmentManager.popBackStack()
+        if(lastFragment is SkillListingFragment){
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt]: Space around paranthesis

File: app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt [hunk 2]
```diff
@@ -64,21 +64,23 @@ class SkillsActivity : AppCompatActivity(), SkillFragmentCallback {
     }
 
     private fun exitActivity(context: Context) {
-        overridePendingTransition(R.anim.trans_right_in, R.anim.trans_right_out)
-        val intent = Intent(context, ChatActivity::class.java)
-        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
-        startActivity(intent)
+        var lastFragment: android.support.v4.app.Fragment? = supportFragmentManager.findFragmentById(R.id.fragment_container)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt]: No nee to cast

File: app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt [hunk 3]
```diff
@@ -64,21 +64,23 @@ class SkillsActivity : AppCompatActivity(), SkillFragmentCallback {
     }
 
     private fun exitActivity(context: Context) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt]: Rename this appropriately, this is not just exitActivity now

File: app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt [hunk 4]
```diff
@@ -64,21 +64,23 @@ class SkillsActivity : AppCompatActivity(), SkillFragmentCallback {
...
         overridePendingTransition(R.anim.trans_right_in, R.anim.trans_right_out)
-        if (supportFragmentManager.popBackStackImmediate(TAG_SKILLS_FRAGMENT, 0)) {
-            title = getString(R.string.skills_activity)
-        } else if (supportFragmentManager.popBackStackImmediate(TAG_GROUP_WISE_SKILLS_FRAGMENT, 0)) {
-            title = getString(R.string.skills_activity)
-        } else {
-            finish()
-            exitActivity(this)
+        supportFragmentManager.addOnBackStackChangedListener {
+            val currentFragment: android.support.v4.app.Fragment? = supportFragmentManager.findFragmentById(R.id.fragment_container)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/SkillsActivity.kt]: No need to cast

File: app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsFragment.kt [hunk 5]
```diff
@@ -123,4 +123,8 @@ class GroupWiseSkillsFragment : Fragment(), IGroupWiseSkillsView, SwipeRefreshLa
         groupWiseSkillsPresenter.onDetach()
         super.onDestroyView()
     }
+    override fun onResume() {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/groupwiseskills/GroupWiseSkillsFragment.kt]: Newline above

**Comment by NileshPranami** [comment]
@iamareebjamal  Sir you are correct about the popBackStack but if onBackPressed() method is overridden then the popBackStack method must be implemented in that. Otherwise that will not work. Please suggest me for alternate of exitActivity after then I will push a final request. Thank you sir

**Comment by NileshPranami** [comment]
@iamareebjamal  Sir, super method really worked. Thanks for your suggestion. Once again review new changes.

**Comment by NileshPranami** [comment]
@iamareebjamal Sir actually #1789 is dependent on this PR so can I add that's code in this or that will for another PR

**Comment by iamareebjamal** [comment]
Great work

**Comment by iamareebjamal** [comment]
@NileshPranami Please follow commit and PR guidelines from now on. See the PR title and description now

**Comment by NileshPranami** [comment]
@iamareebjamal Sure sir, next time trying not to do mistakes. Thanks a lot sir.

**Pull request merged**